core/src/consensus.rs

```diff
@@ -787,14 +787,79 @@ impl Tower {
         false
     }
 
-    fn is_candidate_slot_descendant_of_last_vote(
+    /// Checks if a vote for `candidate_slot` is usable in a switching proof
+    /// to a fork with a common ancestor `common_ancestor` from `last_voted_slot`
+    ///
+    /// Returns None if `candidate_slot` is not present in `ancestors`
+    fn is_valid_switching_proof_vote(
         candidate_slot: Slot,
         last_voted_slot: Slot,
+        common_ancestor: Slot,
+        ancestors: &HashMap<Slot, HashSet<Slot>>,
+    ) -> Option<bool> {
+        trace!("Checking if {candidate_slot} is a valid switching proof vote from {last_voted_slot} with ancestor {common_ancestor}");
+        // Ignore if the `candidate_slot` is a descendant of the `last_voted_slot`, since we do not
+        // want to count votes on the same fork.
+        if Self::is_descendant_slot(candidate_slot, last_voted_slot, ancestors)? {
+            return Some(false);
+        }
+
+        // Only consider forks that split at `common_ancestor` or earlier. This is to prevent
+        // situations like this from being included in the switching proof:
+        //
+        //         /-- `last_voted_slot`
+        //     /--Y
+        //    X    \-- `candidate_slot`
+        //     \-- `switch_slot`
+        //
+        // The `common_ancestor` of `last_voted_slot` and `switch_slot` is `X`. Votes for the `candidate_slot`
+        // should not count towards the switch proof since `candidate_slot` is "on the same fork" as `last_voted_slot`
+        // in relation to `switch_slot`.
+        // However these candidate slots should be allowed:
+        //
+        //             /-- Y -- `last_voted_slot`
+        //    V - W - X
+        //        \    \-- Z -- `switch_slot` -- `candidate_slot`
+        //         \    \-- `candidate_slot`
+        //          \-- `candidate_slot`
+        //
+        // As the `candidate_slot`s forked off from `X` or earlier.
+        //
+        // To differentiate, we check the common ancestor of `last_voted_slot` and `candidate_slot`.
+        // If the common ancestor is <= `common_ancestor` (X in the diagram), we know that it can be counted.
+        Self::greatest_common_ancestor(ancestors, candidate_slot, last_voted_slot)
+            .map(|ancestor| ancestor <= common_ancestor)
```

Comment:
might be clearer/more efficient to check if the switch slot is descended from the common ancestor of candidate_slot, last_voted slot here via: `ancestors.get(switch_slot).contains(ancestor)`. skips the extra `greatest_common_ancestor` call from outside this function to get `common_ancestor`